mantidproject/mantid
# Non axis aligned cutting tool in sliceviewer

**Description of work.**

Added new tool for non axis aligned cuts - see roadmap link in issue for more detailed requirements.

**To test:**
(0) Repeat the instructions in PR #33362 - (there was a  regression of this bug fix that has been re-implemented in this PR)
(1) Check the cutting tool button is only enabled for 3D MD workspaces where all dimensions are Q by opening the following workspaces in sliceviewer. It should only be enabled for the `ws_3D` and `ws_3D_QLab` workspaces (see comment for details) - the first 3 column headers for the vectors should be a*,b*,c* and Qx,Qy,Qz for the two workspaces respectively.

![image](https://user-images.githubusercontent.com/55979119/165113746-37cc3fed-2acb-4d7e-960b-120d67f0d316.png)
```
Load(Filename='CNCS_7860_event.nxs', OutputWorkspace='CNCS_7860_event')  # disabled (MatrixWorkspace)

ws_2D = CreateMDWorkspace(Dimensions='2', Extents='-5,5,-4,4', Names='H,K',
                          Units='r.l.u.,r.l.u.', Frames='HKL,HKL',
                          SplitInto='2', SplitThreshold='50')  #  disabled (2D MD)

ws_3D = CreateMDWorkspace(Dimensions='3', Extents='-5,5,-4,4,-3,3',
                          Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                          Frames='HKL,HKL,HKL', SplitInto='2', SplitThreshold='50')  # enabled!
    
ws_3D_nonQdim = CreateMDWorkspace(Dimensions=3, Extents=[-1, 1, -1, 1, -1, 1], 
                          Names="E,H,K", Frames='General Frame,HKL,HKL', 
                          Units='meV,r.l.u.,r.l.u.')  # disabled (3D but 1 non-Q)
                          
ws_4D = CreateMDWorkspace(Dimensions=4, Extents=[-1, 1, -1, 1, -1, 1, -1, 1], 
                          Names="E,H,K,L", Frames='General Frame,HKL,HKL,HKL', 
                          Units='meV,r.l.u.,r.l.u.,r.l.u.')  # disabled (4D - one non-Q)

ws_3D_QLab = CreateMDWorkspace(Dimensions='3', Extents='-5,5,-4,4,-3,3',
                               Names='Q_lab_x,Q_lab_y,Q_lab_z', Units='U,U,U',
                               Frames='QLab,QLab,QLab', SplitInto='2', SplitThreshold='50')  # enabled!
```
(2) Close any sliceviewer windows and clear the workspaces in the ADS
(3) Run the following and open `ws` in sliceviewer.
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-5,5,-4,4,-3,3',
                          Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                          Frames='HKL,HKL,HKL', SplitInto='2', SplitThreshold='50') 
expt_info = CreateSampleWorkspace()
SetUB(expt_info, 1,1,2,90,90,120)
ws.addExperimentInfo(expt_info)
# make some fake data
FakeMDEventData(ws, UniformParams='1e5', PeakParams='1e+05,0,0,1,0.3', RandomSeed='3873875')
```
(4) Click on non-ortho view of H and K axes - this should disable the cutting tool button
(5) Turn off non-ortho view and check that opening the cut tool:
- Disables non-ortho view
- Disables ROI tool
- Disables line plots

It should look like this
![image](https://user-images.githubusercontent.com/55979119/165118214-39255136-6a2e-4b08-a219-b3bb22f73722.png)

(6) Check that transposing the axes (X <-> Y) will swap the u1 and u2 vectors in the table
(7) Set the axes to (X,Y) = (L, K) - check u1 = [0,0,1] and u2 = [0,1,0]
(8) Change the slice point of H to be 0 - it should look like this
![image](https://user-images.githubusercontent.com/55979119/165118671-58c5aead-c527-484b-969d-db07e303b1cb.png)
(9) In the table double the step along u1 (i.e. set step to 0.12) - this should change nbins = 25 along u1
(10) Set the nbins = 50 along u1 - the step should go back to it's original value (0.06)
(11) Set the stop for u1 to 0 and check that 
- step size = 0.03
- the cut representation line on the colorfill plot has the correct start/stop
- the 1D plot in the cut viewer pane has the correct axes limits

(12) Set step of u1 to be 2 (i.e. greater than the extent of the cut) - this should set nbins=1 and step = 1.5 and put the cut along u2 with nbins = 50.
(13) transpose the axes so now (X,Y) = (K,L) - the cut should have 50 bins along K (default value)
(14) Change the nbins of u2 to 50 (it should set nbins=1 for u1 and change the step=4). Check the white line of the cut representation on the colorfill plot is now vertical.
(15) Try to change the a* column of the u1 to 1 (this would take u1 out of the plane of the slice, i.e. not orthogonal to u3) - it should reset to 0 - i.e. u1 = [0,1,0].
(16) Click and hold down on one of the red markers with white face on the colorfill plot and drag, release at ~K=1.
- This should reset the vectors int he table such that the cut is along u1 = [0,0,1] - i.e. u1 <-> u2
- The thickness along u2 should be adjusted to ~2

(17) Set the step of u2 = 2 in the table, check that it sets (start,stop) = (-1,1)
(18) Drag the top white marker of the cut representation up to L~2
- u1 ~ [0,0,-1]  and u2 ~ [0,-1,0]
- There should be a peak in the 1D plot at x~-1

(19) For u1 change c* = -1 and b* = 0 - check that u2 = [0,-1,0]
(20) To change the centre of the cut move the central white marker of the cut representation to (K,L) ~ (2,0),
- The entire cut representation should move
- The axes label of the 1D plot should be ~`(0.0, 2.0,0.0-1.0x) in 3.14 Ang^-1`
- There should be no peak on the 1D plot

(21) Increase the thickness by dragging the left red marker of the cut representation to encompass the peak in the data - check the peak appears in the 1D plot at the right thickness.
(22) Play around with the direction of the cut by dragging the white markers at the end points of the white line - the vectors u1 and u3 should be orthogonal unit vectors.
(24) Reset the cut by transposing the axes so (X,Y) = (L,K)
(25) Double the slice thickness along H from 0.1 -> 0.2 (the counts of the peak in the 1D plot should double from ~2000 -> ~4000)
(26) Change c* of u1 from 1 -> -1
- The peak in the 1D plot should move from x= 1 -> -1
- Check u2 = [0,-1,0]


Fixes #33631






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
